nixos/modules/services/web-apps/node-red.nix

```diff
@@ -0,0 +1,151 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.services.node-red;
+  defaultUser = "node-red";
+in
+{
+  options.services.node-red = {
+    enable = mkEnableOption "the Node-RED service";
+
+    package = mkOption {
+      default = pkgs.nodePackages.node-red;
+      defaultText = "pkgs.nodePackages.node-red";
+      type = types.package;
+      description = "Node-RED package to use.";
+    };
+
+    openFirewall = mkOption {
+      type = types.bool;
+      default = false;
+      description = ''
+        Open ports in the firewall for the server.
+      '';
+    };
+
+    withNpmAndGcc = mkOption {
```

Comment:
*Maybe* we replace this option with an `example` (and `description`) in the `package` option?
```
options.services.node-red.package = mkOption {
  default = pkgs.nodePackages.node-red;
  defaultText = "pkgs.nodePackages.node-red";
  type = types.package;
  description = ''
    Node-RED package to use.

    If you wish to give Node-RED access to NPM and GCC at runtime, so Nodes can be downloaded and managed
    via the 'Pallet Manager' you can use the example provided. Note that this will blah blah blah imperativeagainst
    blah blah blah work against the "nix way", etc...
  '';
  example = literalExample ''
    pkgs.nodePackages.node-red.override (oldArgs: {
      nativeBuildInputs = [ pkgs.makeWrapper];
      buildInputs = oldArgs.buildInputs ++ [ pkgs.nodePackages.npm pkgs.gcc ];
      postInstall = ''
        wrapProgram "$out/bin/node-red" --prefix PATH : ${
          pkgs.lib.makeBinPath [ pkgs.nodejs pkgs.gcc ]
        }
      '';
    })
  '';
};
```

Having this as an actual option seems like NixOS is promoting something that NixOS doesn't want to promote, so including a simple example seems more fitting in my humble opinion.

Reply:
I disagree. This service is only useful with access to the library of 3000+ nodes available from https://flows.nodered.org/search?type=node.I think it's very bad to provide services that are not functional or practical at all. I would only support the addition of Node-RED as a service to nixpkgs with this option remaining present.

e5bc2a3 makes the `withNpmAndGcc` option wrap any `cfg.package` 
4ab3d42 updates the description in line with what you're suggesting, but only as much as necessary.

Reply:
Unless you plan to add all 3000 nodes yourself to Nixpkgs and keep them up to date, I don't really see an alternative to allowing this option to exist. Otherwise, the service should not be in Nix at all, and I'll just provide it as a flake, if Nixpkgs is really this hostile to functionality.

Reply:
If I'm honest, this option should be set to true, since you won't really be able to do that much with Node-RED without access to the palette library.

Reply:
> This service is only useful with access to the library of 3000+ nodes available from https://flows.nodered.org/search?type=node.

Thanks for the context. It sounds like the option doesn't make sense then... because it should always be true?

Reply:
> Thanks for the context. It sounds like the option doesn't make sense then... because it should always be true?

No. Because someone could write an expression to include the nixpkgs using node2nix, or any number of methods of declaring the packages to be 100% declared.

In such a case, it would be very relevant to remove NPM and GCC, because if Node-RED detects that NPM/GCC are not present at runtime, it disables the internal palette manager.

This option exists for the majority of people, not the few that will express every out-of-nixpkgs npm dependency they want.

Reply:
yep, until we can get ecosystem like node red to become nix way friendly, then stuff like can't be run natively on nix. A chicken and egg problem. Maybe we should lay and egg, make chicken, then ask the chicken to lay egg's differently.

Reply:
> yep, until we can get ecosystem like node red to become nix way friendly, then stuff like can't be run natively on nix. A chicken and egg problem. Maybe we should lay the egg, make chicken, then ask the chicken to lay egg's differently.

It runs fine on Nix. What does native mean? There is no way we're ever going to incorporate other package ecosystems 100% into expressions, the problem is the underlying software. We cannot support something that wants to use a different package ecosystem, using the Nix language. All we can do is give those pieces of software access to whatever pieces of software they want at runtime, so they can do their thing.

Reply:
We can automate some of that work. But that's a long way away, and I'd rather get software into nixpkgs instead of eternal bikeshedding about how declarative something is, as if it's the job of the software to change right now, just because we want to package it.